### ISSUE TITLE
s2s-ft中实现细节的疑惑

### ISSUE THREAD

**Issue by hxs91** [microsoft/unilm#581]
**Describe**
Model I am using (s2s-ft):
您好，我对s2s-ft中三种mask策略实现细节有一些疑惑，描述如下：

v0:
	1. 共用sep和sos，会不会破坏sep的语义？
	2. 在设置mask_way为v0时，如果不设置target_mask_prob为0，会出现大量p(token_t|(..., token_t-2, [MASK]))的情况，这样的微调好像和论文里的描述有出入？

v1:
	好像和论文里的描述没有出入~

v2:
	不太理解为啥生成pseudo_id的时候也是使用和v1同样的mask策略（以target_mask_prob的概率mask，并且按照1:1:8的概率分配mask_token），这里如果全都使用pseudo_id([MASK])会有什么坏处呢？

**Comment by addf400** [comment]
@hxs91 感谢关注
v0是指 [UniLM](https://arxiv.org/abs/1905.03197) 的fine-tuning方式，区别与v1的Causal方式（用上一个token的表示去预测下一个token）这里是需要对被预测的token进行mask然后才能计算对应的loss。即只在被mask的位置出loss。
v2中设置target_mask_prob 能够提高下游任务上的性能，类似word drop的效果。

**Comment by hxs91** [comment]
@addf400 谢谢回复，但还是没解决我的疑惑。
我看了UniLM的fine-tuning方式，我理解那应该是s2s-ft中的v1。v0中如果target_mask_prob不设置为0（默认值为0.5）的话，
> 
    if r < self.target_mask_prob and pos > 0:
          target_ids[pos] = self.get_masked_token(target_ids[pos])

根据上述代码，会有50%的p(token_t|(..., token_t-2, [MASK]))，这里丢掉了token_t-1，我理解您说的mask的位置出loss，但mask位置并不是由target_ids决定的，

> 
    def gather_seq_out_by_pos(seq, pos):
      return torch.gather(seq, 1, pos.unsqueeze(2).expand(-1, -1, seq.size(-1)))

    sequence_output = outputs[0]
    target_sequence_output = sequence_output[:, source_len:,]
    masked_sequence_output = gather_seq_out_by_pos(target_sequence_output,
                                                   masked_pos)

    def loss_mask_and_normalize(loss, mask):
      mask = mask.type_as(loss)
      loss = loss * mask
      denominator = torch.sum(mask) + 1e-5
      return (loss / denominator).sum()

    prediction_scores_masked = self.cls(masked_sequence_output)

可以看到loss实际是由masked_pos决定的。所以是不是这里其实和v2的实现一样，也是为了提高下游任务的性能（word drop）？主要看到v0和v2在实现上这些细节论文里没有提到才有此疑惑。。

**Comment by hxs91** [comment]
v2里的这种实现，在每个pseudo_token计算loss的时候和v1是一样的（因为context里都有50%的词被mask掉），不同点只在于，v2里的每个token都会去计算一次loss。不知道这样理解对吗？

**Comment by addf400** [comment]
> (v0: position shift, v1: masked LM, v2: pseudo-masking)



> @addf400 谢谢回复，但还是没解决我的疑惑。 我看了UniLM的fine-tuning方式，我理解那应该是s2s-ft中的v1。v0中如果target_mask_prob不设置为0（默认值为0.5）的话，
> 
> > 
> 
> ```
> if r < self.target_mask_prob and pos > 0:
>       target_ids[pos] = self.get_masked_token(target_ids[pos])
> ```
> 
> 根据上述代码，会有50%的p(token_t|(..., token_t-2, [MASK]))，这里丢掉了token_t-1，我理解您说的mask的位置出loss，但mask位置并不是由target_ids决定的，
> 
> > 
> 
> ```
> def gather_seq_out_by_pos(seq, pos):
>   return torch.gather(seq, 1, pos.unsqueeze(2).expand(-1, -1, seq.size(-1)))
> 
> sequence_output = outputs[0]
> target_sequence_output = sequence_output[:, source_len:,]
> masked_sequence_output = gather_seq_out_by_pos(target_sequence_output,
>                                                masked_pos)
> 
> def loss_mask_and_normalize(loss, mask):
>   mask = mask.type_as(loss)
>   loss = loss * mask
>   denominator = torch.sum(mask) + 1e-5
>   return (loss / denominator).sum()
> 
> prediction_scores_masked = self.cls(masked_sequence_output)
> ```
> 
> 可以看到loss实际是由masked_pos决定的。所以是不是这里其实和v2的实现一样，也是为了提高下游任务的性能（word drop）？主要看到v0和v2在实现上这些细节论文里没有提到才有此疑惑。。

谢谢指出笔误! (v0: position shift, v1: masked LM, v2: pseudo-masking)

1) 在fine-tuning使用 v0 和 v2的时候，设置--target_mask_prob 的目的都是一样的，功能类似word drop
2) v1 只能用被mask 掉的 token去计算loss （全被mask掉了就没有context），v2 因为有Pseudo token，可以在所有token上计算loss

**Comment by hxs91** [comment]
@addf400 嗯嗯，感谢，明白了。

**Comment by addf400** [comment]
(v0: position shift, v1: masked LM)
v0 与 v1的fine-tuning是共享modeling部分的
具体的区别就是model的input的构造，具体参考代码：
https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L115
https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L136

**Comment by hxs91** [comment]
> (v0: position shift, v1: masked LM) v0 与 v1的fine-tuning是共享modeling部分的 具体的区别就是model的input的构造，具体参考代码： https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L115 https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L136

这里是了解的，其实我如果仔细读一读README应该也能get到word dropout。不过说到这，还有个疑问，就是我在最开始提的，实现里是共用[SOS]和[SEP]的，这样做不会有什么影响吗？

**Comment by addf400** [comment]
> > (v0: position shift, v1: masked LM) v0 与 v1的fine-tuning是共享modeling部分的 具体的区别就是model的input的构造，具体参考代码： https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L115 https://github.com/microsoft/unilm/blob/master/s2s-ft/s2s_ft/utils.py#L136
> 
> @addf400 共用modeling这里我是了解的，其实我如果仔细读一读README应该也能get到word dropout。不过说到这，还有个疑问，就是我在最开始提的，实现里是共用[SOS]和[SEP]的，这样做不会有什么影响吗？

s2s-ft 里面共用[SOS]和[SEP]是一个可行的方案，但是并没有尝试其他的策略。
